Realm/RLMObjectStore.mm

```diff
@@ -95,45 +95,178 @@ void RLMInitializeSwiftAccessorGenerics(__unsafe_unretained RLMObjectBase *const
+static void validateValueForProperty(__unsafe_unretained id const obj,
+                                        __unsafe_unretained RLMProperty *const prop) {
+    switch (prop.type) {
+        case RLMPropertyTypeString:
+        case RLMPropertyTypeBool:
+        case RLMPropertyTypeDate:
+        case RLMPropertyTypeInt:
+        case RLMPropertyTypeFloat:
+        case RLMPropertyTypeDouble:
+        case RLMPropertyTypeData:
+            if (!RLMIsObjectValidForProperty(obj, prop)) {
+                @throw RLMException(@"Invalid value '%@' for property '%@'", obj, prop.name);
+            }
+            break;
+        case RLMPropertyTypeObject:
+            break;
+        case RLMPropertyTypeArray: {
+            if (obj != nil && obj != NSNull.null) {
+                if (![obj conformsToProtocol:@protocol(NSFastEnumeration)]) {
+                    @throw RLMException(@"Array property value (%@) is not enumerable.", obj);
+                }
+            }
+            break;
         }
+        case RLMPropertyTypeAny:
+        case RLMPropertyTypeLinkingObjects:
+            @throw RLMException(@"Invalid value '%@' for property '%@'", obj, prop.name);
     }
+}
 
-    // if no existing, create row
-    created = NO;
-    if (rowIndex == realm::not_found) {
-        try {
-            rowIndex = table.add_empty_row();
-        }
-        catch (std::exception const& e) {
-            @throw RLMException(e);
+static NSUInteger createRowForObject(RLMClassInfo const& info) {
+    try {
+        return info.table()->add_empty_row();
+    }
+    catch (std::exception const& e) {
+        @throw RLMException(e);
+    }
+}
+
+/* If a row exists with the specified primary key value, return its index. Otherwise, return `realm::not_found`.
+ *
+ * Precondition: `info` must refer to a class which has a primary key property
+ * Precondition: `primaryValue` is a validated property value that has been coerced to `nil`
+ */
+static NSUInteger getRowForObjectWithPrimaryKey(RLMClassInfo const& info, id primaryValue) {
+    REALM_ASSERT_DEBUG(info.propertyForPrimaryKey());
+
+    RLMProperty *const primaryProperty = info.propertyForPrimaryKey();
+    const NSUInteger primaryPropertyColumn = info.tableColumn(primaryProperty);
+
+    try {
+        switch (primaryProperty.type) {
+            case RLMPropertyTypeString:
+                return info.table()->find_first_string(primaryPropertyColumn, RLMStringDataWithNSString(primaryValue));
+
+            case RLMPropertyTypeInt:
+                if (primaryValue) {
+                    return info.table()->find_first_int(primaryPropertyColumn, [primaryValue longLongValue]);
+                } else {
+                    return info.table()->find_first_null(primaryPropertyColumn);
+                }
+
+            default:
+                REALM_UNREACHABLE();
         }
-        created = YES;
+
+    }
+    catch (std::exception const& e) {
```

Comment:
What exceptions is this catching? I don't think any of the things called here can throw.